### ISSUE TITLE
OpenHAB2 generates weak SSL keys

### ISSUE THREAD

**Issue by xsnrg** [openhab/openhab2#271]
When connecting to port 8443 with the latest Firefox (38.0.1) which was updated to reject weak keys by Firefox, the following message is displayed and there is no ability to proceed:

Secure Connection Failed

An error occurred during a connection to <site>:8443. SSL received a weak ephemeral Diffie-Hellman key in Server Key Exchange handshake message. (Error code: ssl_error_weak_server_ephemeral_dh_key)

**Comment by kaikreuzer** [comment]
Any idea how to change this in the Jetty setup?

**Comment by Stratehm** [comment]
Two solutions:
-Increase the size of the ephemeral Diffie-Hellman key size to 1024 or above (Add the system property jdk.tls.ephemeralDHKeySize to 1024 or 2048 before jetty startup)
-Disable all SSL/TLS cipher suites that use an ephemeral Diffie-Hellman key. All cipher suites that starts with DHE or EDH. After a sslscan on OpenHAB, only disable EDH-RSA-DES-CBC3-SHA and DHE-RSA-AES128-SHA SSL cipher suites in the jetty-ssl.xml configuration file of the bundle.

The second solution should be prefered but may prevent some old browsers to use SSL.

**Comment by kaikreuzer** [comment]
It seems that option 1 requires Java 8 (https://docs.oracle.com/javase/8/docs/technotes/guides/security/enhancements-8.html) is this right?
If so, I would be fine with option 2. Would you be able to create a PR for it?

**Comment by Stratehm** [comment]
Indeed, first solution is for Java8 only.

PR #277 created to disable problematic cipher suites.

**Comment by kaikreuzer** [comment]
Thanks a lot!

**Comment by xsnrg** [comment]
Verified.  Thanks!

**Comment by teichsta** [comment]
the only error i got (on demo.openhab.org:9443 which is on 2.0.0b) is:

```
demo.openhab.org:9443 uses an invalid security certificate.
The certificate is not trusted because it is self-signed.
The certificate is only valid for openhab.org
(Error code: sec_error_unknown_issuer) 
```

which is the same for openhab 1 on demo.openhab.org:8443.

Could you also help to apply the necessary changes for OH1 as well?

Thanks, Thomas E.-E.

**Comment by Stratehm** [comment]
This warning has not the same cause. It says that the certificate is a self-signed certificate:
Each SSL certificate has to be signed by a private key of a Certificate Authority which is trusted by your browser (like Comodo, GlobalSign...). This certification is not free and has to be done for each domain name you want to secure and authenticate. 
Since OpenHAB is mainly installed on private server, a new certificate has to be generated for each installation. Since servers are private, authentication of the server is not really useful and only ciphering is wanted.
For such case, the new certificate can be self-signed, i.e. signed with its own private key (and thus that is not trusted by default by browsers). The warning says that the server you are attempting to connect cannot be trusted. 

In conclusion, this warning cannot be easily removed but is not really important on private server and since ciphering is not impacted, your connection is secure. (Except that you may be exposed to some Man-in-the-middle attack, but it is another story).

I hope it is clear enough ;)

**Comment by bartcerneels** [comment]
Will/can this be backported to the 1.x version?
